keras_nlp/models/llama/llama_attention.py

```diff
@@ -0,0 +1,194 @@
+        rope_scaling_type=None,
+        rope_scaling_factor=None,
+        dropout=0.0,
+        rotary_percentage=0.25,
+        kernel_initializer="glorot_uniform",
+        bias_initializer="zeros",
+        rotary_max_wavelength=10000,
+        max_sequence_length=512,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.hidden_dim = hidden_dim
+        self.num_heads = num_heads
+        self.num_key_value_heads = num_key_value_heads
+        self.num_key_value_groups = num_key_value_groups
+        self.attn_head_size = hidden_dim // num_heads
+        self.rotary_max_wavelength = rotary_max_wavelength
+
+        self.kernel_initializer = keras.initializers.get(kernel_initializer)
+        self.bias_initializer = keras.initializers.get(bias_initializer)
+        self.max_sequence_length = max_sequence_length
+
+        self.rotary_dim = int(self.attn_head_size * rotary_percentage)
+        self.rope_scaling_type = rope_scaling_type
+        self.rope_scaling_factor = rope_scaling_factor
+
+        if rope_scaling_type is None:
+            self.rotary_embedding_layer = RotaryEmbedding(
+                max_wavelength=rotary_max_wavelength
+            )
+        else:
+            if rope_scaling_type == "linear":
+                self.rotary_embedding_layer = RotaryEmbedding(
+                    max_wavelength=rotary_max_wavelength
+                )
+            elif rope_scaling_type == "dynamic":
+                self.rotary_embedding_layer = DynamicNTKScaledRotaryEmbedding(
+                    max_wavelength=rotary_max_wavelength
+                )
+            else:
+                raise ValueError(
+                    f"Unknown RoPE scaling type {rope_scaling_type}"
+                )
+
+    def build(self, input_shape):
+        self._query_dense = keras.layers.EinsumDense(
+            equation="abc,cde->abde",
+            output_shape=(None, self.num_heads * self.attn_head_size),
+            name="query",
+        )
+        self._query_dense.build(input_shape)
+
+        self._key_dense = keras.layers.EinsumDense(
+            equation="abc,cde->abde",
+            output_shape=(None, self.num_key_value_heads * self.attn_head_size),
+            name="key",
+        )
+        self._key_dense.build(input_shape)
+
+        self._value_dense = keras.layers.EinsumDense(
+            equation="abc,cde->abde",
+            output_shape=(None, self.num_key_value_heads * self.attn_head_size),
+            name="value",
+        )
+        self._value_dense.build(input_shape)
+
+        self._softmax = keras.layers.Softmax(axis=-1, name="attention_softmax")
+
+        self._output_dense = keras.layers.EinsumDense(
+            equation="abc,cd->abd",
+            output_shape=(None, self.hidden_dim),
+            name="attention_output",
+        )
+        self._output_dense.build(input_shape)
+
+        self.built = True
+
+    def call(self, hidden_states, attention_mask=None):
```

Comment:
https://github.com/huggingface/transformers/blob/41c5f45bfe958be58bac6c891652f632ebac23e2/src/transformers/models/llama/modeling_llama.py#L292-L307

HF implements this forward block by splitting weights and calling F.linear, 
1. What is `config.pretraining_tp` ?
2. How should we implement this?

Reply:
> config.pretraining_tp

See this issue https://github.com/huggingface/transformers/pull/24906, this looks like it is a tricky issue with numerical accuracy to fully emulate a partitioned dense call during pre-training. I would just fully ignore for now, we could play with this down the road when we are checking numerical accuracy of checkpoints.

Reply:
Yup. That's why I chucked this for now because this would need scoping after we match checkpoints (which is a bigger challenge in itself because of Llama size).